runtime/compiler/runtime/RuntimeAssumptions.cpp

```diff
@@ -51,15 +51,6 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::make(
    return result;
    }
 
-
-void
-TR_PatchNOPedGuardSiteOnClassPreInitialize::compensate(TR_FrontEnd *fe, bool isSMP, void *data)
-   {
-   reclaim();
-   TR::PatchNOPedGuardSite::compensate(fe, isSMP, data);
-   }
```

Comment:
@andrewcraik @dsouzai it was not easily possible to leave this code in place. I assume the original intention of this code was to reclaim the persistent memory allocated for the class signature right after calling compensate as this is the earliest such a call can be made. However calling `reclaim` at this point makes things difficult for us because we are now zeroing out the `_key` after freeing the memory to ensure we don't double free.

Because the `_key` is zero following a call to `compensate` we can no longer calculate the hashcode of this assumption by calling the `hashcode` API [1], which is called from [2] after calling `compensate`. i.e. to mark the correct bucket in the RAT which needs to be reclaimed later we must calculate the hashcode [3], and to calculate the hashcode we need the class signature which we already freed.

The alternatives are to precalculate the hashcode and stash it off on the assumption object aggressively or to not free the class signature presistent memory and let the reclamation logic take care of it, which should happen shortly after compensating anyway.

In this PR I chose the second option as it is simpler and without any surprises. It is done at the cost of a very slight time window in which a small amount of persistent memory may be alive when it doesn't need to be. I think the complexity in this case is not worth the trade off for the space savings. However I'd like your opinion on this.

[1] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/RuntimeAssumptions.cpp#L63-L91
[2] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/RuntimeAssumptions.cpp#L366
[3] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/ClassUnloadAssumption.cpp#L376-L383

Reply:
With the lazy reclamation the window for it to be freed is not guaranteed to be very small, but I think the more important thing is to fix the hangs and other issues definitively with a 'simple ' change and we can then try to optimize the memory management. Perhaps we could free up a bit in the hashcode with a left shift and the pointer field could point to the string or if low tagged could be interpreted as a hashcode left in place of the string. More complex for sure, but would probably achieve what we want in terms of memory management etc.

That being said, I am fine with this fix for now.

Reply:
The way I see it, the allocated string in `_key` is essentially part of the runtime assumption. It is already the case that assumptions in general are not immediately reclaimed after they fire. Therefore, we shouldn't treat the string allocation any different. By reclaiming all aspects of a given runtime assumption in one place at one time, it simplifies the code and prevents non-trivial bugs. 

tl;dr I'm fine with the fix as it stands.

Reply:
I am ok too with the solution chosen here.